### ISSUE TITLE
Add “consistent-as-needed” option for “quotes” (especially when “avoidEscape” is on)

### ISSUE THREAD

**Issue by sunyatasattva** [eslint/eslint#14693]
**What rule do you want to change?**

`quotes` rule

**Does this change cause the rule to produce more or fewer warnings?**

More warnings.

**How will the change be implemented? (New option, new default behavior, etc.)?**

The rule `quote-props` has the `consistent-as-needed` option, which tells the user that if any of an object literal props require quotes, then all props should be wrapped in quotes.

Similarly, the `quotes` rule has the `avoidEscape` option, which allows the user to use a different quote style in case that would allow for fewer escaped characters. This, however, results in potentially very unpleasant object literals (especially, in TS-speak, `Record<string, string>` kind) with inconsistent quote styles within one literal.

I am aware of a potential conflict of priorities: what if there are multiple values within the object literal that would need different quote styles depending on what needs less escaping? In this case, there are a few implementation options:

1. The user decides which rule should take prio
2. Both are allowed and consistency is not enforced in that case

I believe a similar problem is however already present with the `avoidEscape`, e.g. a string that looks like: `It's time to go "home"` needs to be escaped no matter which quote style is used (excluding backticks of course), and I think the current behavior gives freedom to the user to choose, so I think it would be consistent to implement this option as (2.) above.

**Please provide some example code that this change will affect:**

```js
// Assume the rule is quotes: ["error", "double", { avoidEscape: true }]
const selectors = {
  myButton: '[title="my-button"]',
  myDiv: "#my-div",
  // ...etc
}
```

**What does the rule currently do for this code?**
This code would pass the linting under the current rule.

**What will the rule do after it's changed?**
This code will need to use consistent quote styles within the object:

```js
// Assume the rule is quotes: ["error", "double", { avoidEscape: true }]
const selectors = {
  myButton: '[title="my-button"]',
  myDiv: '#my-div',
  // ...etc
}
```

**Are you willing to submit a pull request to implement this change?**

Yes.

**Comment by nzakas** [comment]
I’m sorry, we are no longer making changes to rules that simply enforce stylistic preferences. (See the [README](https://github.com/eslint/eslint#stylistic-rule-updates) for details.)
